graph/src/blockchain/firehose_block_ingestor.rs

```diff
@@ -0,0 +1,156 @@
+                .await;
+
+            match result {
+                Ok(stream) => latest_cursor = self.process_blocks(latest_cursor, stream).await,
+                Err(e) => {
+                    error!(self.logger, "Unable to connect to endpoint: {:?}", e);
+                }
+            }
+
+            // If we reach this point, we must wait a bit before retrying
+            backoff.sleep_async().await;
+        }
+    }
+
+    async fn fetch_head_cursor(&self) -> String {
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+        loop {
+            match self.chain_store.clone().chain_head_cursor() {
+                Ok(cursor) => return cursor.unwrap_or_else(|| "".to_string()),
+                Err(e) => {
+                    error!(self.logger, "Fetching chain head cursor failed: {:?}", e);
+
+                    backoff.sleep_async().await;
+                }
+            }
+        }
+    }
+
+    async fn process_blocks(
+        &self,
+        cursor: String,
+        mut stream: Streaming<bstream::BlockResponseV2>,
+    ) -> String {
+        let mut latest_cursor = cursor;
+
+        while let Some(message) = stream.next().await {
+            match message {
+                Ok(v) => {
+                    let step = bstream::ForkStep::from_i32(v.step)
+                        .expect("Fork step should always match to known value");
+
+                    let result = match step {
+                        bstream::ForkStep::StepNew => self.process_new_block(&v).await,
+                        bstream::ForkStep::StepUndo => {
+                            trace!(self.logger, "Received undo block to ingest, skipping");
+                            Ok(())
+                        }
+                        _ => panic!(
+                            "We explicitely requested StepNew|StepUndo but received something else"
+                        ),
+                    };
+
+                    if let Err(e) = result {
+                        error!(self.logger, "Process block failed: {:?}", e);
+                        break;
+                    }
+
+                    latest_cursor = v.cursor;
+                }
+                Err(e) => {
+                    info!(
+                        self.logger,
+                        "An error occurred while streaming blocks: {}", e
+                    );
+                    break;
+                }
+            }
+        }
+
+        error!(
+            self.logger,
+            "Stream blocks complete unexpectedly, expecting stream to always stream blocks"
+        );
+        latest_cursor
+    }
+
+    async fn process_new_block(&self, response: &bstream::BlockResponseV2) -> Result<(), Error> {
+        let block = decode_firehose_block::<M>(response)
+            .context("Mapping firehose block to blockchain::Block")?;
+
+        trace!(self.logger, "Received new block to ingest {}", block.ptr());
```

Comment:
Trace is off by default and this isn't that frequent, lets make it `debug!`.

Reply:
This is much more frequent on higher throughput chain like NEAR, printed each second. I would keep it at trace personally. What do you think?

Reply:
Once a second isn't so bad, and in setups with multiple graph node processes the ingestor is its own process. That said, the ideal thing might be to keep this as trace but add a prometheus metric to track the latest block ingested, like the `ethereum_chain_head_number` metric in the current block ingestor.

Reply:
Ok I'll create a new issue once we have this PR merged to do that (metric of `firehose_chain_head_number`).